flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,268 @@
+        return getProject().files((Object[]) jars);
+    }
+
+    @Input
+    List<String> getAgentArgs() {
+        return javaAgents.stream().map(it -> it.args).collect(Collectors.toList());
+    }
+
+    public void javaAgent(Action<JavaAgent> action) {
+        JavaAgent agent = new JavaAgent();
+        action.execute(agent);
+        if (agent.jar == null) throw new GradleException("No jar file specified for Java agent");
+        javaAgents.add(agent);
+    }
+
+    @Inject
+    public FlaskJarTask(ObjectFactory objects) {
+        BasePluginConvention basePluginConvention = getProject().getConvention().getPlugin(BasePluginConvention.class);
+        getDestinationDirectory().set(basePluginConvention.getLibsDir());
+        getArchiveBaseName().convention(getProject().getName());
+        getArchiveExtension().convention("jar");
+        launcherClassName = objects.property(String.class).convention(Flask.Constants.DEFAULT_LAUNCHER_NAME);
+        mainClassName = objects.property(String.class);
+        jvmArgs = objects.listProperty(String.class);
+        javaAgents = new ArrayList<>();
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
+
+        Provider<File> heartbeatJarProvider = getProject().provider(() -> {
+            File dest = new File(FlaskJarTask.this.getTemporaryDir(), HeartbeatAgentResource.instance.getDisplayName());
+            try (OutputStream os = new FileOutputStream(dest); InputStream is = HeartbeatAgentResource.instance.read()) {
+                Flask.write2Stream(is, os);
+            }
+            return dest;
+        });
+        includeLibraries(heartbeatJarProvider);
+    }
+
+    @Input
+    @SneakyThrows
+    public String getLauncherArchiveHash() {
+        return Flask.bytes2Hex(Flask.computeSHA256Digest(LauncherResource.instance::read));
+    }
+
+    @Input
+    @SneakyThrows
+    public String getHeartbeatAgentHash() {
+        return Flask.bytes2Hex(Flask.computeSHA256Digest(HeartbeatAgentResource.instance::read));
+    }
+
+    @RequiredArgsConstructor
+    private static class StreamAction implements CopyActionProcessingStreamAction {
+
+        private final ZipOutputStream zoos;
+        private final Manifest manifest;
+        private final MessageDigest md;
+        private final byte[] buffer;
+
+        @Override
+        @SneakyThrows
+        public void processFile(FileCopyDetailsInternal fileCopyDetails) {
+            String entryName = fileCopyDetails.getRelativePath().toString();
+            if (!fileCopyDetails.isDirectory() && entryName.startsWith(Flask.Constants.LIBRARIES_FOLDER)) {
+                Supplier<InputStream> streamSupplier = new Supplier<InputStream>() {
+                    @Override
+                    @SneakyThrows
+                    public InputStream get() {
+                        return new FileInputStream(fileCopyDetails.getFile());
+                    }
+                };
```

Comment:
Lambda power!
```java
Supplier<InputStream> streamSupplier = () -> new FileInputStream(fileCopyDetails.getFile());
```

Reply:
this is the thing I hate the most about Java... you can't use a lambda here because `FileInputStream` constructor throws a checked exception while `Supplier#get` doesn't have any `throws` declaration, so you either have to catch it, wrap into a `RuntimeException` and rethrow it or use `@SneakyThrows`, which is an annotation, and you cannot annotate lambda in Java 😡

Reply:
Well, I _was_ going to suggest creating a function like this: :grin:
```java
private static InputStream createInputStream(File file) {
    try {
        return new FileInputStream(file);
    } catch (IOException e) {
        throw new UncheckedIOException(e);
    }
}
```

Reply:
lol... I wasn't aware of the existence of `UncheckedIOException`.. sounds like they finally accepted the fact that checked exceptions are a mistake

Reply:
btw, I'll do that, it is certainly more concise